hexa/workspaces/schema/queries.py

```diff
@@ -62,6 +65,44 @@ def resolve_pending_workspace_invitations(_, info, page=1, per_page=10):
     return result_page(qs, page=page, per_page=per_page)
 
 
+@workspace_queries.field("workspaceCandidates")
+def resolve_users(_, info, **kwargs):
+    request = info.context["request"]
+    query = kwargs["query"].strip()
+    workspaceSlug = kwargs["workspace"].strip()
+
+    try:
+        workspace = Workspace.objects.get(slug=workspaceSlug)
+        members = WorkspaceMembership.objects.all().filter(workspace=workspace)
+
+        is_admin = (
+            members.filter(user=request.user).first().role
+            == WorkspaceMembershipRole.ADMIN
+        )
+
+        if not is_admin:
+            return {
+                "items": [],
+            }
+
+        queryset = User.objects.all()
+
+        if query is not None:
+            queryset = (
+                queryset.filter(email__icontains=query)
+                .exclude(email__in=members.values_list("user__email", flat=True))
+                .order_by("email")[:10:1]
+            )
+
+        return {
+            "items": queryset,
+        }
+    except Exception:
```

Comment:
I'd prefer to catch the exceptions we know could happen (`Workspace.DoesNotExist`, `WorkspaceMembership.DoesNotExist`) to return an empty list and raise an error otherwise. Hiding errors this way could come back to haunt us in the future :p

Reply:
In fact, here is how I would have done it (with a few more years of Python :p) 

```python
@workspace_queries.field("workspaceCandidates")
def resolve_users(_, info, **kwargs):
    request = info.context["request"]
    query = kwargs["query"].strip()

    try:
        workspace = Workspace.objects.filter_for_user(request.user).get(slug=kwargs["workspaceSlug"])

        if not request.user.has_perm("workspaces.manage_members", workspace):
            raise PermissionDenied

        # Find users by looking in email, last name or first name for a case-insensitive match
        qs = User.objects.filter(
            Q(email__icontains=query)
            | Q(last_name__icontains=query)
            | Q(first_name__icontains=query)
        )

        # Exclude users that are already members of the workspace
        qs = qs.exclude(id__in=workspace.members.values_list("id", flat=True))

        return qs.order_by("email")[:10]
    except PermissionDenied:
        return []
    except Workspace.DoesNotExist:
        return []


```